### DISCUSSION TITLE
How to parse a mustache like syntax?

### DISCUSSION THREAD

**Discussion by ashokgelal** [petitparser/dart-petitparser#65] [Q&A] [answered]
Sorry for asking what could be a trivial question but how'd one start parsing a mustache like syntax? Something like:

```html
<h2>{{ foo }} </h2>
{{ bar }}
<p> text here </p>
```
Where the mustache tags `{{ foo }}` and `{{ bar }}` are optional.

I sort of got it working but my solution feels like a total hack (I'm embarrassed to even show it). I'm sure there must be a better way to parse something like this. I'm using `ref()`s if that matters.

**Comment by renggli** [comment, marked as answer]
Not sure how exactly you want to parse your example, but the following parser separates the text and the mustache parts nicely:

```dart
final startMustache = string('{{');
final endMustache = string('}}');

final mustache =
    startMustache & any().starLazy(endMustache).flatten() & endMustache;
final content = startMustache.neg().plus().flatten();

final document = (mustache | content).star().end();

final input = '''<h2>{{ foo }} </h2>
{{ bar }}
<p> text here </p>''';
final output = document.parse(input);
print(output);
```

Not sure what you use `ref()` for? This is normally only useful when you have a recursive grammar. In this example I don't see recursion, but maybe I misunderstood what you are trying to do?

**Comment by ashokgelal** [comment]
Thank you! I was close but not close enough. `lazy` and `greedy` look like very useful functions but that are not well documented. It'd be very helpful for beginners if there were couple of examples on how to use them.

As far as `ref()` is concerned, my real grammar is recursive but I dumbed down the example in my question to make it simpler.
